File: ios/AccessEye/Camera/CameraManager.swift
//
//  CameraManager.swift
//  AccessEye
//
//  Thin AVFoundation wrapper: runs a live capture session and grabs ONE still
//  frame on demand (tap-to-capture). The session is configured and driven on a
//  dedicated queue; @Published state is always updated on the main actor so
//  SwiftUI stays happy. (README §6.1, M1.)
//
//  Declared `nonisolated` so the capture-delegate callbacks and session work can
//  run off the main actor; UI-facing state is pushed back to main explicitly.
//

import AVFoundation
import UIKit
import Combine

nonisolated final class CameraManager: NSObject, ObservableObject {

    /// The session SwiftUI's preview layer renders.
    let session = AVCaptureSession()

    /// Whether camera access was denied/restricted, so the UI can explain it.
    @Published var permissionDenied = false

    /// Whether the session is configured and running.
    @Published var isRunning = false

    private let sessionQueue = DispatchQueue(label: "gr.orestislef.AccessEye.camera")
    private let photoOutput = AVCapturePhotoOutput()
    private var isConfigured = false

    /// Completion for the in-flight capture, if any. Touched only on sessionQueue.
    private var captureCompletion: ((UIImage?) -> Void)?

    // MARK: - Lifecycle

    /// Request permission (if needed), configure once, and start the preview.
    func start() {
        switch AVCaptureDevice.authorizationStatus(for: .video) {
        case .authorized:
            configureAndRun()
        case .notDetermined:
            AVCaptureDevice.requestAccess(for: .video) { [weak self] granted in
                guard let self else { return }
                if granted {
                    self.configureAndRun()
                } else {
                    self.setPermissionDenied(true)
                }
            }
        case .denied, .restricted:
            setPermissionDenied(true)
        @unknown default:
            setPermissionDenied(true)
        }
    }

    /// Stop the preview (e.g. when leaving the screen).
    func stop() {
        sessionQueue.async { [weak self] in
            guard let self, self.session.isRunning else { return }
            self.session.stopRunning()
            self.setRunning(false)
        }
    }

    // MARK: - Capture

    /// Capture a single still frame. `completion` is called on the main actor.
    func capturePhoto(completion: @escaping (UIImage?) -> Void) {
        sessionQueue.async { [weak self] in
            guard let self, self.session.isRunning else {
                Task { @MainActor in completion(nil) }
                return
            }
            self.captureCompletion = completion
            let settings = AVCapturePhotoSettings()
            self.photoOutput.capturePhoto(with: settings, delegate: self)
        }
    }

    // MARK: - Internals

    private func configureAndRun() {
        sessionQueue.async { [weak self] in
            guard let self else { return }
            if !self.isConfigured {
                self.configureSession()
            }
            guard self.isConfigured, !self.session.isRunning else { return }
            self.session.startRunning()
            self.setRunning(self.session.isRunning)
        }
    }

    /// Build the session graph: back wide camera input → photo output.
    private func configureSession() {
        session.beginConfiguration()
        session.sessionPreset = .photo

        guard
            let device = AVCaptureDevice.default(.builtInWideAngleCamera,
                                                 for: .video, position: .back),
            let input = try? AVCaptureDeviceInput(device: device),
            session.canAddInput(input)
        else {
            session.commitConfiguration()
            setPermissionDenied(true)
            return
        }
        session.addInput(input)

        guard session.canAddOutput(photoOutput) else {
            session.commitConfiguration()
            return
        }
        session.addOutput(photoOutput)

        session.commitConfiguration()
        isConfigured = true
    }

    private func setPermissionDenied(_ value: Bool) {
        Task { @MainActor in self.permissionDenied = value }
    }

    private func setRunning(_ value: Bool) {
        Task { @MainActor in self.isRunning = value }
    }
}

// MARK: - AVCapturePhotoCaptureDelegate

extension CameraManager: AVCapturePhotoCaptureDelegate {
    func photoOutput(_ output: AVCapturePhotoOutput,
                     didFinishProcessingPhoto photo: AVCapturePhoto,
                     error: Error?) {
        let completion = captureCompletion
        captureCompletion = nil

        let image: UIImage? = {
            guard error == nil,
                  let data = photo.fileDataRepresentation(),
                  let img = UIImage(data: data) else { return nil }
            return img
        }()

        Task { @MainActor in completion?(image) }
    }
}

File: android/app/src/main/java/gr/orestislef/accesseye/camera/CameraManager.kt
//
// CameraManager.kt
// AccessEye
//
// Thin CameraX wrapper: runs a live preview and grabs ONE still frame on demand
// (tap-to-describe). Port of Camera/CameraManager.swift. All camera work happens
// off the main thread (CameraX executors + coroutines); UI-facing state is
// exposed as StateFlows so Compose stays happy.
//
// Unlike iOS, the runtime permission REQUEST lives in the UI layer
// (rememberLauncherForActivityResult) — the result is fed back here via
// onPermissionResult(granted), which mirrors the iOS requestAccess callback.
//

package gr.orestislef.accesseye.camera

import android.Manifest
import android.content.Context
import android.content.pm.PackageManager
import android.graphics.Bitmap
import android.graphics.Matrix
import androidx.camera.core.CameraSelector
import androidx.camera.core.ImageCapture
import androidx.camera.core.ImageCaptureException
import androidx.camera.core.ImageProxy
import androidx.camera.core.Preview
import androidx.camera.lifecycle.ProcessCameraProvider
import androidx.camera.lifecycle.awaitInstance
import androidx.core.content.ContextCompat
import androidx.lifecycle.LifecycleOwner
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.suspendCancellableCoroutine
import kotlinx.coroutines.withContext
import kotlin.coroutines.resume

class CameraManager(private val context: Context) {

    /// Whether camera access was denied, so the UI can explain it (spoken + shown).
    private val _permissionDenied = MutableStateFlow(false)
    val permissionDenied: StateFlow<Boolean> = _permissionDenied.asStateFlow()

    /// Whether the use cases are bound and the preview is live.
    private val _isRunning = MutableStateFlow(false)
    val isRunning: StateFlow<Boolean> = _isRunning.asStateFlow()

    /// The Preview use case CameraPreview attaches its PreviewView surface to.
    /// Created eagerly so the composable can attach before/after binding freely.
    val preview: Preview = Preview.Builder().build()

    /// Still capture tuned for speed over quality — a blind user waiting for a
    /// description cares about latency, and Gemma downscales the image anyway.
    private val imageCapture: ImageCapture = ImageCapture.Builder()
        .setCaptureMode(ImageCapture.CAPTURE_MODE_MINIMIZE_LATENCY)
        .build()

    // Binding must happen on the main thread; keep our own scope so start/stop
    // don't depend on the caller's.
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate)
    private var bindJob: Job? = null
    private var cameraProvider: ProcessCameraProvider? = null

    /// The lifecycle we should bind to once permission arrives (start() may run
    /// before the permission dialog is answered).
    private var pendingOwner: LifecycleOwner? = null

    private val hasPermission: Boolean
        get() = ContextCompat.checkSelfPermission(context, Manifest.permission.CAMERA) ==
            PackageManager.PERMISSION_GRANTED

    // MARK: - Lifecycle

    /// Bind preview + still capture to the back camera if permission is already
    /// granted; otherwise remember the owner and wait for onPermissionResult.
    fun start(lifecycleOwner: LifecycleOwner) {
        pendingOwner = lifecycleOwner
        if (hasPermission) {
            _permissionDenied.value = false
            bind(lifecycleOwner)
        }
        // else: the UI layer is requesting permission; onPermissionResult(true)
        // will bind with the owner we just stored.
    }

    /// Result of the UI-layer permission request (mirrors iOS requestAccess).
    fun onPermissionResult(granted: Boolean) {
        _permissionDenied.value = !granted
        if (granted) pendingOwner?.let { bind(it) }
    }

    /// Stop the preview (e.g. when leaving the screen). Safe to call repeatedly.
    fun stop() {
        bindJob?.cancel()
        bindJob = null
        pendingOwner = null
        cameraProvider?.unbindAll()
        _isRunning.value = false
    }

    // MARK: - Capture

    /// Capture a single still frame, rotation applied, or null on any failure.
    /// Guard: capturing while the camera is not running returns null (mirrors iOS).
    suspend fun capturePhoto(): Bitmap? {
        if (!_isRunning.value) return null
        val image = awaitStill() ?: return null
        // Decode + rotate off the main thread; frames can be large.
        return withContext(Dispatchers.Default) {
            try {
                image.use { proxy ->
                    val degrees = proxy.imageInfo.rotationDegrees
                    val bitmap = proxy.toBitmap()
                    if (degrees == 0) bitmap
                    else Bitmap.createBitmap(
                        bitmap, 0, 0, bitmap.width, bitmap.height,
                        Matrix().apply { postRotate(degrees.toFloat()) }, true,
                    )
                }
            } catch (e: CancellationException) {
                throw e
            } catch (_: Exception) {
                null // failure surfaces as a spoken retry upstream
            }
        }
    }

    // MARK: - Internals

    /// Wrap ImageCapture's callback API in a coroutine. Never throws — a failed
    /// capture resumes with null so callers get the same "no photo" path as iOS.
    private suspend fun awaitStill(): ImageProxy? = suspendCancellableCoroutine { cont ->
        imageCapture.takePicture(
            ContextCompat.getMainExecutor(context),
            object : ImageCapture.OnImageCapturedCallback() {
                override fun onCaptureSuccess(image: ImageProxy) {
                    if (cont.isActive) cont.resume(image) else image.close()
                }

                override fun onError(exception: ImageCaptureException) {
                    if (cont.isActive) cont.resume(null)
                }
            },
        )
    }

    /// Get the process-wide provider and (re)bind our use cases to the back camera.
    private fun bind(owner: LifecycleOwner) {
        bindJob?.cancel()
        bindJob = scope.launch {
            try {
                val provider = ProcessCameraProvider.awaitInstance(context)
                cameraProvider = provider
                provider.unbindAll()
                provider.bindToLifecycle(
                    owner,
                    CameraSelector.DEFAULT_BACK_CAMERA,
                    preview,
                    imageCapture,
                )
                _isRunning.value = true
            } catch (e: CancellationException) {
                throw e
            } catch (_: Exception) {
                // No usable back camera / bind failure: stay not-running so the
                // capture guard returns null and the UI speaks a retry.
                _isRunning.value = false
            }
        }
    }
}
